### PR TITLE
FOGL-3604:  upgrade/downgrade

### PR DESCRIPTION
**Pull request by sstefano69**
Signed-off-by: Stefano Simonelli <stefano@dianomic.com>

SQLlite

1.7.0.RC

image

upgrade:

image

downgrade:
image

image

PG:
Init:
image

downgrade:

image

upgrade:

image

image

### CODE REVIEW

File: scripts/plugins/storage/postgres/upgrade/32.sql [hunk 1]
```diff
@@ -0,0 +1,41 @@
+--
+-- DROP the column: read_key   uuid
+--
+
+ALTER TABLE readings DROP CONSTRAINT readings_pkey;
```
Comment (ashish-jabble) [inline scripts/plugins/storage/postgres/upgrade/32.sql]:
```
ALTER TABLE readings DROP COLUMN IF EXISTS read_key;
```
I believe PostgreSQL will automatically remove all of its indexes and constraints involving the column

**Review by ashish-jabble — CHANGES_REQUESTED** [review]

File: scripts/plugins/storage/postgres/upgrade/32.sql (hunk 1 above)
Comment (sstefano69) [reply scripts/plugins/storage/postgres/upgrade/32.sql]: fixed

Comment (sstefano69) [reply scripts/plugins/storage/postgres/upgrade/32.sql]: @ashish-jabble , I encountered a problem with the SQLIte downgrade that I didn't see before, I have fixed it, it should work now.

Comment (ashish-jabble) [reply scripts/plugins/storage/postgres/upgrade/32.sql]: yeah before it was failing on sqlite downgrade.
```
Failure in downgrade command [/usr/bin/sqlite3 '/usr/local/foglamp/data/foglamp.db' "ATTACH DATABASE '/usr/local/foglamp/data/foglamp.db' AS 'foglamp'; .read '/usr/local/foglamp/scripts/plugins/storage/sqlite/downgrade/31.sql' .quit"]: Error: near line 15: UNIQUE constraint failed: new_readings.read_key. Exiting
```
Now with latest changes it is working fine.

**Comment by ashish-jabble** [comment]
@sstefano69 I see some issues with downgrade postgres sql script. i.e **new_readings_read_key_key** constraint is NOT deleting

downgrade

File: scripts/plugins/storage/postgres/downgrade/31.sql [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    ts
+FROM readings;
+
+DROP INDEX fki_readings_fk1;
+DROP INDEX readings_ix2;
+DROP INDEX readings_ix3;
+
+DROP TABLE readings;
+
+ALTER TABLE new_readings rename to readings;
```
Comment (ashish-jabble) [inline scripts/plugins/storage/postgres/downgrade/31.sql]: Cannot we do it with ADD COLUMN? Like given below queries
```
ALTER TABLE foglamp.readings ADD COLUMN read_key uuid UNIQUE;
UPDATE foglamp.readings SET read_key = uuid_in(md5(random()::text || clock_timestamp()::text)::cstring);
```
Also there is no use of drop & re-creating indexes with given approach.

Comment (ashish-jabble) [inline scripts/plugins/storage/postgres/downgrade/31.sql]: Cannot we do it with ADD COLUMN? Like given below queries
```
ALTER TABLE foglamp.readings ADD COLUMN read_key uuid UNIQUE;
UPDATE foglamp.readings SET read_key = uuid_in(md5(random()::text || clock_timestamp()::text)::cstring);
```
Also there is no use of drop & re-creating indexes with given approach.

Comment (sstefano69) [reply scripts/plugins/storage/postgres/downgrade/31.sql]: done

**Review by ashish-jabble — APPROVED** [review]

**Pull request merged**